extmod/uasyncio/core.py

```diff
@@ -160,11 +148,8 @@ def run_until_complete(main_task=None):
             if t:
                 # A task waiting on _task_queue; "ph_key" is time to schedule task at
                 dt = max(0, ticks_diff(t.ph_key, ticks()))
-            elif not _io_queue.map:
-                # No tasks can be woken so finished running
-                return
-            # print('(poll {})'.format(dt), len(_io_queue.map))
-            _io_queue.wait_io_event(dt)
+            # print('(poll_ms {})'.format(dt))
+            _io_queue.poll_ms(dt)
```

Comment:
Following up on the above comment about moving mp_handle_pending or MICROPY_EVENT_POLL_HOOK here, doing so would provide more flexibility. I forget what the ph_key of a ready task looks like, so it may require an additional  check when recalculating dt.

Reply:
See reply above about that.  This call should sleep as long as possible.